workflow/controller/taskresult.go

```diff
@@ -33,7 +33,17 @@ func (wfc *WorkflowController) newWorkflowTaskResultInformer() cache.SharedIndex
 			options.LabelSelector = labelSelector
 			// `ResourceVersion=0` does not honor the `limit` in API calls, which results in making significant List calls
 			// without `limit`. For details, see https://github.com/argoproj/argo-workflows/pull/11343
-			options.ResourceVersion = ""
+			// The reflector will record `lastSyncResourceVersion`. If `ResourceVersion != "0"`, we should use the value
+			// recorded by the reflector. see https://github.com/argoproj/argo-workflows/pull/13466
+			if options.ResourceVersion == "0" {
+				options.ResourceVersion = ""
+			}
+			// The reflector will set the Limit to `0` when `ResourceVersion != "" && ResourceVersion != "0"`, which will fail
+			// to limit the number of workflow returns. Timeouts and other errors may occur when there are a lots of workflows.
+			// see https://github.com/kubernetes/client-go/blob/ee1a5aaf793a9ace9c433f5fb26a19058ed5f37c/tools/cache/reflector.go#L286
+			if options.Limit == 0 {
+				options.Limit = common.DefaultPageSize
+			}
```

Comment:
I'm thinking we may want to split this whole block into a common helper function that we use for all these informers so that it's consolidated and consistent and harder to accidentally change one and not another

Reply:
Resolved, I added two new util methods.

Reply:
@agilgur5 Could you take a view again, thanks!